sstables/sstables_manager.cc

```diff
@@ -149,8 +149,40 @@ sstable_writer_config sstables_manager::configure_writer(sstring origin) const {
     return cfg;
 }
 
-void sstables_manager::update_total_memory_used_by_components(const sstable* sst) {
+void sstables_manager::update_total_memory_used_by_components_and_maybe_reclaim(const sstable* sst) {
     _total_memory_used_by_components += sst->total_components_memory_size();
+
+    if (_total_memory_used_by_components <= _max_allowed_memory_for_components) {
+        // total memory used is within limit; no need to reclaim.
+        return;
+    }
+
+    // Memory consumption has crossed threshold; Some of it needs to be reclaimed.
+    // Sort the _active list in descending order of total memory, so that we can
+    // start reclaiming from the largest sstable in terms of memory.
+    _active.sort([] (const sstable& sst1, const sstable& sst2) {
+        return sst1.total_components_memory_size() > sst2.total_components_memory_size();
+    });
+
+    // Reclaim memory until the total usage gets under limit.
+    auto prev_total_memory_used_by_components = _total_memory_used_by_components;
+    for (auto& sst : _active) {
+        if (sst.total_components_memory_size() == 0) {
+            // No more memory to reclaim from this SSTable or any remaining SSTables in the
+            // list as they are sorted in descending order of total_components_memory_size().
+            break;
+        }
+
+        _total_memory_used_by_components -= sst.reclaim_memory_from_components();
+
+        if (_total_memory_used_by_components < _max_allowed_memory_for_components) {
+            // total memory usage is now under limit
+            smlogger.info("Reclaimed {} bytes of memory from SSTable components", prev_total_memory_used_by_components - _total_memory_used_by_components);
+            return;
+        }
+    }
+
+    smlogger.warn("Reclaimed {} bytes of memory from SSTable components, but it is insufficient to meet the threshold", prev_total_memory_used_by_components - _total_memory_used_by_components);
```

Comment:
This can never happen currently because we only account the reclaimable components in the limit and so it is always possible to reduce the amount to 0, by reclaiming all components.

Reply:
I agree. In addition to this, the new version reclaims from only one SSTable. So I have removed the warning.